File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        int n = nums.size();
        
        int p=1;
        
        int maxi=INT_MIN;
        
        for(int i=0; i<n; i++){
            
            p *= nums[i];
            maxi = max(p, maxi);
            
            if(nums[i]==0) p=1;
        }
        
        p=1;
        for(int i=n-1; i>=0; i--){
            
            p *= nums[i];
            maxi = max(maxi, p);
            
            if(nums[i]==0) p=1;
        }
        
        return maxi;
    }
};

File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        res, cur_product = nums[0], 1
        
        for num in nums:
            cur_product *= num
            res = max(res, cur_product)
            if cur_product == 0:
                cur_product = 1
        
        cur_product = 1
        for i in range(len(nums)-1, -1, -1):
            cur_product *= nums[i]
            res = max(res, cur_product)
            if cur_product == 0:
                cur_product = 1
        return res
    
    